Code/src/main/java/nl/utwente/group10/haskell/expr/Function.java

```diff
@@ -0,0 +1,124 @@
+package nl.utwente.group10.haskell.expr;
+
+import nl.utwente.group10.haskell.env.Env;
+import nl.utwente.group10.haskell.exceptions.HaskellException;
+import nl.utwente.group10.haskell.hindley.GenSet;
+import nl.utwente.group10.haskell.hindley.HindleyMilner;
+import nl.utwente.group10.haskell.type.FuncT;
+import nl.utwente.group10.haskell.type.Type;
+
+import java.util.*;
+
+/**
+ * Represents a user-defined Haskell function. Holds arguments which can be used for building the function.
+ */
+public class Function extends Expr {
+
+    /**
+     * An argument for a function. A FunctionArgument can be used as a variable in a function definition.
+     */
+    public static class FunctionArgument extends Expr {
+        /** The type for this argument. */
+        private Type type;
+
+        /** The name of this argument. Will be randomly generated to a unique value. */
+        private String name;
+
+        /**
+         * Constructor for a FunctionArgument with an explicit type.
```

Comment:
What does 'explicit' mean here? `type` could still be a `new VarT()`, not a lot of explicitness about that.

Could you require `type` to be a tree of constant types perhaps?